### PR TITLE
LoginLogOut First Commit

### PR DESCRIPTION
**Pull request by gravidesaldrin**
This is the controllers, models, helpers, views and routes of the login logut task. I will send the other files for my next PR. Please review, thanks.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      log_in @user
+      flash[:success] = t ".title"
+      redirect_to @user
+    else
+      render "new"
+    end
+  end
+
+  def edit
+    @user = User.find params[:id]
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: you can use `before_action` to avoid DRY code

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  def destroy
+    User.find(params[:id]).destroy
+    flash[:success] = t ".title"
+    redirect_to users_url
+  end
+
+  private
+
+    def user_params
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: indent = 2 spaces

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      params.require(:user).permit(:name, :email, :password,
+                                   :password_confirmation)
+    end
+
+    def logged_in_user
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: indent = 2 spaces

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      unless logged_in?
+        store_location
+        redirect_to login_url
+      end
+    end
+
+    def correct_user
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: indent = 2 spaces

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      @user = User.find(params[:id])
+      redirect_to root_url unless current_user? @user
+    end
+
+    def admin_user
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: indent = 2 spaces

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    def logged_in_user
+      unless logged_in?
+        store_location
+        redirect_to login_url
+      end
+    end
+
+    def correct_user
+      @user = User.find(params[:id])
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: remove `()`

File: app/models/user.rb [hunk 7]
```diff
@@ -1,12 +1,47 @@
...
+  validates :email, presence: true, length: {maximum: 255},
+                    format: {with: VALID_EMAIL_REGEX},
+                    uniqueness: {case_sensitive: false}
+  validates :password, presence: true, length: {minimum: 6}, allow_nil: true
+
+  before_save {self.email = email.downcase}
+
+  has_secure_password
+
+  def User.digest string
```
Comment (dieunb) [inline app/models/user.rb]: You neen read convention again to implement class method

File: app/models/user.rb [hunk 8]
```diff
@@ -1,12 +1,47 @@
...
+    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST :
+                                                  BCrypt::Engine.cost
+    BCrypt::Password.create string, cost: cost
+  end
+  def User.new_token
```
Comment (dieunb) [inline app/models/user.rb]: You neen read convention again to implement class method.
Add new white line above this

File: app/models/user.rb [hunk 9]
```diff
@@ -1,12 +1,47 @@
...
+    SecureRandom.urlsafe_base64
+  end
+  def remember
```
Comment (dieunb) [inline app/models/user.rb]: Add new white line above this

File: app/models/user.rb [hunk 10]
```diff
@@ -1,12 +1,47 @@
...
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
```
Comment (dieunb) [inline app/models/user.rb]:
```
update_attributes :remember_digest, User.digest(remember_token)
```

Should compare with `update_attribute`

File: app/models/user.rb [hunk 11]
```diff
@@ -1,12 +1,47 @@
...
+  end
+  def authenticated? remember_token
```
Comment (dieunb) [inline app/models/user.rb]: Add new white line above this

File: app/models/user.rb [hunk 12]
```diff
@@ -1,12 +1,47 @@
...
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+  def forget
```
Comment (dieunb) [inline app/models/user.rb]: Add new white line above this

File: app/models/user.rb [hunk 13]
```diff
@@ -1,12 +1,47 @@
...
+    update_attribute :remember_digest, nil
```
Comment (dieunb) [inline app/models/user.rb]:
```
update_attributes :remember_digest, User.digest(remember_token)
```

Should compare with `update_attribute`

File: app/views/sessions/new.html.erb [hunk 14]
```diff
@@ -0,0 +1,26 @@
+<% provide :title, t(".title") %>
+<h1>Log in</h1>
```
Comment (dieunb) [inline app/views/sessions/new.html.erb]: i18n

File: app/views/sessions/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+<div class="row">
+  <div class="col-lg-8">
+    <%= form_for :session, url: login_path,
+                html: {class: "margin-bottom-40"} do |f| %>
```
Comment (dieunb) [inline app/views/sessions/new.html.erb]: indent

File: app/views/sessions/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+      </div>
+      <div class = "checkbox">
+        <%= f.label :remember_me, class: "checkbox inline" do %>
+          <%= f.check_box :remember_me %>
+          <span> <%= t ".remember_me" %> </span>
+        <% end %>
+      </div>
+      <%= f.submit t(".title"), class: "btn btn-primary" %>
+    <% end %>
+    <p> New user? <%= link_to t(".sign_up_now"), signup_path %> </p>
```
Comment (dieunb) [inline app/views/sessions/new.html.erb]: i18n

File: app/views/shared_pages/_error_messages.html.erb [hunk 17]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(object.errors.count, "error") %>.
```
Comment (dieunb) [inline app/views/shared_pages/_error_messages.html.erb]: i18n

File: app/views/shared_pages/_navbar.html.erb [hunk 18]
```diff
@@ -12,19 +12,43 @@
...
-        <li>
-          <%= link_to "#" do %>
-            <i class="glyphicon glyphicon-lock"></i>
-            <%= t ".log_in" %>
-          <% end %>
-        </li>
+        <% if logged_in? %>
+          <li class="dropdown">
+            <%= link_to "#", class: "dropdown-toggle", role: "button",
+                  "data-toggle": "dropdown" do %>
```
Comment (dieunb) [inline app/views/shared_pages/_navbar.html.erb]: indent

File: app/views/shared_pages/_navbar.html.erb [hunk 19]
```diff
@@ -12,19 +12,43 @@
...
+              <i class="glyphicon glyphicon-user"></i>
+              <%= t ".user" %>
+            <% end %>
+
+            <ul id="g-account-menu" class="dropdown-menu" role="menu">
+                <li><%= link_to t(".my_profile"), current_user %> </li>
```
Comment (dieunb) [inline app/views/shared_pages/_navbar.html.erb]: indent

File: app/views/shared_pages/_navbar.html.erb [hunk 20]
```diff
@@ -12,19 +12,43 @@
...
+                <li>
```
Comment (dieunb) [inline app/views/shared_pages/_navbar.html.erb]: indent

File: app/controllers/sessions_controller.rb [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
+      redirect_back_or user
+    else
+      render "new"
```
Comment (MuhammadTamzid) [inline app/controllers/sessions_controller.rb]: use symbol

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+  end
+
+  def create
+    @user = User.new user_params
+    if @user.save
+      log_in @user
+      flash[:success] = t ".title"
+      redirect_to @user
+    else
+      render "new"
```
Comment (MuhammadTamzid) [inline app/controllers/users_controller.rb]: use symbol

File: app/controllers/users_controller.rb [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+    @user = User.find params[:id]
+  end
+
+  def update
+    @user = User.find params[:id]
+    if @user.update_attributes user_params
+      flash[:success] = t ".title"
+      redirect_to @user
+    else
+      render "edit"
```
Comment (MuhammadTamzid) [inline app/controllers/users_controller.rb]: use symbol

File: app/controllers/users_controller.rb [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    User.find(params[:id]).destroy
+    flash[:success] = t ".title"
+    redirect_to users_url
+  end
+
+  private
+
+    def user_params
+      params.require(:user).permit(:name, :email, :password,
+                                   :password_confirmation)
```
Comment (MuhammadTamzid) [inline app/controllers/users_controller.rb]: indent

File: app/models/user.rb [hunk 25]
```diff
@@ -1,12 +1,47 @@
...
+  end
+
```
Comment (MuhammadTamzid) [inline app/models/user.rb]: remove blank line

File: app/views/shared_pages/_error_messages.html.erb (hunk 17 above)
Comment (MuhammadTamzid) [inline app/views/shared_pages/_error_messages.html.erb]: i18n

File: app/views/shared_pages/_navbar.html.erb (hunk 19 above)
Comment (MuhammadTamzid) [inline app/views/shared_pages/_navbar.html.erb]: indent, two space

File: app/views/shared_pages/_error_messages.html.erb [hunk 26]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      <%= t ".intro" %> <%= pluralize(object.errors.count, "error") %>.
```
Comment (MuhammadTamzid) [inline app/views/shared_pages/_error_messages.html.erb]: remove()

File: app/controllers/users_controller.rb [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+  def correct_user
+    @user = User.find params[:id]
+    redirect_to root_url unless current_user? @user
+  end
+
+  def admin_user
+    redirect_to root_url unless current_user.admin?
+  end
+
+  def get_user
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: `load_user` or `find_user`, don't use `get`

**Comment by gravidesaldrin** [comment]
Updated Sir. I change it to find_user. OK.

File: app/controllers/users_controller.rb [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+  def update
+    if @user.update_attributes user_params
+      flash[:success] = t ".title"
+      redirect_to @user
+    else
+      render :edit
+    end
+  end
+
+  def destroy
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: we will destroy user on admin namspace.

**Comment by gravidesaldrin** [comment]
Ok SIr. I remove the destroy action for this commit.

File: app/controllers/users_controller.rb [hunk 29]
```diff
@@ -0,0 +1,56 @@
+class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:edit, :update]
+  before_action :correct_user, only: [:edit, :update]
+  before_action :find_user
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: For all actions?

**Comment by gravidesaldrin** [comment]
Ok Sir. Updated. Put :except for new, create

**Pull request merged**